fix(mui-controlled-form): allow validation objects and forward required

TheSharpieOne opened the pull request:
Fix regression introduced in #666 for using objects for `required` validation (more details about what that means in the original fix: #658)
Fix regression introduced in #666 for passing the `required` property through to underlying `TextField`s

The fixes include supporting the deprecated way of passing `required` directly as well as using the new `rules` object.

Added tests for all of these to prevent future regression.

TheSharpieOne commented on packages/controlled-form/src/lib/AsyncAutocomplete.tsx:
```diff
@@ -66,6 +67,15 @@ export const ControlledAsyncAutocomplete = <
           {...rest}
           FieldProps={{
             ...FieldProps,
+            required: rules?.required
```
What you are saying is for the case of yup/joi/etc, the developer would need to manually set `required` in `FieldProps` for it to appear as required, correct? If that is the case I can move this up to come before `...FieldProps` so that it doesn't override any value manually set.

My goal is to improve DX where possible. Having to specify a field is required in multiple places it tedious and can easily be overlooked resulting in required fields not indicated as required. This is especially the case when using schema based validation. Developers have to generate a schema, then on each individual component specify it is required which defeats the purpose a little (I mean if you have to know it is required when rendering the individual field, then you could have just as well used `rules` on the field.)

In the future to better support something like yup (since I would image that is the most common use-case after `rules` since many apps that migrate will be using it) we can automatically set `required` based on the schema (It is possible with `describe` and can even account for conditionally required `when` cases).
I haven't used joi enough to know if it is possible for that library, but I would imagine it could also be supported if that is a common use-case.

TheSharpieOne replied:
Actually, I already accounted for manually setting `required` in `FieldProps`. It falls back to the value provided in FieldProps. If `required` or `rules.required` is set, it will prefer that. In the case of using a schema based validation (yup/joi/etc) those should not be set.

LauRoxx replied:
TLDR: Can we adjust this so only the `rules.required` gets forwarded to the field instead?

The top level `required` definition is already blurred for the `TextField` & inputs between the old `register` props and the required indicators on the field. I'd prefer to not blur those lines any more across the other controlled fields as it makes it more difficult to see if the prop is being used correctly. Many could see the top level `required` as the field one and since it would add the indicator they'd have no other way to know it's not for that without digging through docs. There is no definition confusion within the `rules` prop, so I can deal with the inconsistency between rules/resolvers with `rules.required` forwarding `required` as a reward for updating to the correct pattern lol. 

This isn't shutting down the idea of bringing more `FieldProps` to the top level, but better for a bit later when there isn't this level of refactor happening.

TheSharpieOne replied:
There was a little added confusion with the deprecation of the older props (such as required) as well as the breaking change. i.e the `required` was deprecated but also it's usage changed in a breaking way.
So we should just rip the bandaid off and remove all of those deprecated props allowing required to just be the boolean it meant to be for the indicator without any additional special logic.

TheSharpieOne commented:
Rebased to [feat/mui-6](https://github.com/Availity/element/tree/feat/mui-6) with the removal of deprecated fields. Updated logic to only set `required` based on `rules.required` which gets overridden by `FieldProps.required` (via spread) if set.
